### PR TITLE
Pipes | Kate Evans-Spitzer and Sara Frandsen | Scrabble

### PR DESCRIPTION
**Pull request by Guribot**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

|  Feature 	|   Feedback	|
|---	|---	|
|   Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? |  Because it makes it more easily accessible to other classes trying to access the method, and the method does not depend on there being an instance of the Scoring class.  |
|   Describe an example of a test you wrote for a _nominal case_. |   "Scoring.highest_score_from - returns a word from the given array". The test passes a hardcoded array of words into the method and checks for the correct result. 	|
|   Describe an example of a test you wrote for an _edge case_. |  "Scoring.highest_score_from - In case of a tie between a 7-letter word and a shorter word, return 7-letter word, regardless of order". Again, the test passes a hardcoded array of words into the method and checks the result. It checks in both orders to confirm the correct criteria is being passed. 	|
|   What was your final test coverage percentage?	|   100%   🎉 💯 🎉 	|
|   Is there a specific piece of code you'd like feedback on?	|   We jumped through some hoops in order to make wave-3-game.rb loop properly (such as returning :fail instead of false if the user tries to play an invalid word, since returning false triggers the player_has_won variable) and are curious as to what an easier way of doing that would be, especially within the confines of the assignment. 	|

### CODE REVIEW

File: lib/scoring.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        elsif score(word) > score(best_word)
+          best_word = word
+        end
+      end
+      best_word
+    end # end of self.highest_score
+
+    def self.handle_tie(word1, word2)
+      if word1.length == 7
+        word1
```
Comment (droberts-sea) [inline lib/scoring.rb:35]: Good job breaking this out into its own method.

File: specs/scoring_spec.rb [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+    it 'adds 50 points for a 7-letter word' do
+      @scrabble.score('jazzily').must_equal 85
+    end
+
+    it 'raise ArgumentError if passed anything but a string' do
+      proc { @scrabble.score(nil) }.must_raise ArgumentError
+      proc { @scrabble.score(true) }.must_raise ArgumentError
+      proc { @scrabble.score(4.0) }.must_raise ArgumentError
+    end
```
Comment (droberts-sea) [inline specs/scoring_spec.rb:38]: What about an empty word, or an 8-letter word?

File: specs/scoring_spec.rb [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+    it 'returns shortest word in case of a tie' do
+      words = %w[jar XD]
+      @scrabble.highest_score_from(words).must_equal 'XD'
+    end
+
+    it 'it returns 7-letter word rather than short word' do
+      words = %w[aeadaea zzzzzj]
+      @scrabble.highest_score_from(words).must_equal 'aeadaea'
+      words = %w[zzzzzj aeadaea]
```
Comment (droberts-sea) [inline specs/scoring_spec.rb:60]: I like that you test both arrangements here. Good work!

File: specs/scoring_spec.rb [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+      @scrabble.highest_score_from(words).must_equal 'aeadaea'
+    end
+
+    it 'returns first word in event of an absolute tie' do
+      words = %w[jazzily ylizzaj]
+      @scrabble.highest_score_from(words).must_equal 'jazzily'
+    end
```
Comment (droberts-sea) [inline specs/scoring_spec.rb:67]: What about an empty list of words, or a list with only one word?

File: specs/player_spec.rb [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+    it 'returns false if player does not have tiles to play word' do
+      @debug_player.play('jazzily').must_equal :fail
+    end
+
+    it 'removes played tiles from hand' do
+      hand1 = @debug_player.tiles.dup
+      @debug_player.play('jazz')
+      @debug_player.tiles.wont_equal hand1
+    end
```
Comment (droberts-sea) [inline specs/player_spec.rb:61]: What happens if you attempt to play after the player has already won?

File: specs/tilebag_spec.rb [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+    it 'draws random tiles' do
+      tiles1 = @test_bag.draw_tiles(7)
+      tiles2 = @test_bag.draw_tiles(7)
+      tiles1.wont_equal tiles2
+    end
+
+    it 'does not draw more tiles than exist' do
+      proc { @test_bag.draw_tiles(99) }.must_raise ArgumentError
+    end
```
Comment (droberts-sea) [inline specs/tilebag_spec.rb:45]: You're missing a couple of interesting edge cases here:
- What happens if you attempt to draw 0 tiles?
- What happens if you attempt to draw the last tile from the bag (but not more than it has)?

**Comment by droberts-sea** [comment]
# Scrabble
## What We're Looking For

|  Feature  |   Feedback |
|---	|---	|
|  **General**	|   	|
|   Answered comprehension questions  |  yes  |
|   Both Teammates contributed to the codebase	|  yes  |
|   Regular Commits with meaningful commit messages  |  yes  |
|  **Scoring class**	|   	|
|  `score` method	|  yes  |
|   Uses appropriate data structure to store the letter score	|  yes  |
|   Appropriately handles edge cases	|  yes  |
|   Tests `score` edge cases (one-letter word, 7 letter word)	|  some - see inline  |
|  `highest_scoring_word` method	|  yes  |
|   Appropriately handles edge cases	|  yes  |
|   Tests `highest_scoring_word` edge cases (ties, empty array)	|  some - see inline  |
|  **Player class**	|   	|
|  Uses an array to keep track of words played	|  yes  |
|  Uses existing code to create `highest_scoring_word` and `highest_word_score`	|  yes  |
|  Returns `true` or `false` from `won?` method |  yes |
|  Tests edge cases on the `play(word)` method for words that cannot be scored |  yes  |
|  **TileBag class**	|   	|
|  Uses a data structure to keep the set of default tiles in the bag	|  yes  |
|  `draw_tiles` method uses the parameter to draw a set of random tiles |  yes  |
| `tiles_remaining` method returns a count	|  yes  |
|  Tests `draw_tiles` edge cases for different parameter inputs	|  some - see inline  |

Good work - this code is well-organized and easy to read!

There are a few places where you've got holes in your test coverage, which I've tried to outline below. Aside from random error cases, the pattern I'm seeing is that you should make sure to explore both sides of any edge cases. For example, in `TileBag#draw_tiles`, while you consider what happens if you attempt to draw more tiles than are in the bag, you don't check that you can draw the last tile in the bag. You can look at your code and say "of course that would work", but until you test it you can't know for sure.

Becoming a master tester is not easy, and this is the first project where you've had to build tests around complex logic, so don't worry if it's not yet where we want it to be. We'll have plenty of time to practice going forward. Do make sure to spend time thinking about it for Hotel, particularly around how ranges of dates interact, and let one of us know if you want us to check your work.
